math/mathcore/src/RanluxppEngineImpl.cxx

```diff
@@ -0,0 +1,209 @@
+ * Copyright (C) 1995-2020, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+/** \class ROOT::Math::RanluxppEngine
+Implementation of the RANLUX++ generator
+
+RANLUX++ is an LCG equivalent of RANLUX using 576 bit numbers.
+
+Described in
+A. Sibidanov, *A revision of the subtract-with-borrow random numbergenerators*,
+*Computer Physics Communications*, 221(2017), 299-303,
+preprint https://arxiv.org/pdf/1705.03123.pdf
+
+The code is loosely based on the Assembly implementation by A. Sibidanov
+available at https://github.com/sibidanov/ranluxpp/.
+*/
+
+#include "Math/RanluxppEngine.h"
+
+#include "mulmod.h"
+
+#include <cstdint>
+
+namespace {
+
+// Variable templates are a feature of C++14, use the older technique of having
+// a static member in a template class.
+
+template <int p>
+struct RanluxppData;
+
+template <>
+struct RanluxppData<1> {
+   static const uint64_t kA[9];
+};
+const uint64_t RanluxppData<1>::kA[] = {
+   0x0000000000000001, 0x0000000000000000, 0x0000000000000000, 0xffff000001000000, 0xffffffffffffffff,
+   0xffffffffffffffff, 0xffffffffffffffff, 0xffffffffffffffff, 0xfffffeffffffffff,
+};
+
+template <>
+struct RanluxppData<24> {
+   static const uint64_t kA[9];
+};
+const uint64_t RanluxppData<24>::kA[] = {
+   0x0000000000000000, 0x0000000000000000, 0x0000000000010000, 0xfffe000000000000, 0xffffffffffffffff,
+   0xffffffffffffffff, 0xffffffffffffffff, 0xfffffffeffffffff, 0xffffffffffffffff,
+};
+
+template <>
+struct RanluxppData<2048> {
+   static const uint64_t kA[9];
+};
+const uint64_t RanluxppData<2048>::kA[] = {
+   0xed7faa90747aaad9, 0x4cec2c78af55c101, 0xe64dcb31c48228ec, 0x6d8a15a13bee7cb0, 0x20b2ca60cb78c509,
+   0x256c3d3c662ea36c, 0xff74e54107684ed2, 0x492edfcc0cc8e753, 0xb48c187cf5b22097,
+};
+
+} // end anonymous namespace
+
+namespace ROOT {
+namespace Math {
+
+template <int w, int p>
+class RanluxppEngineImpl {
+
+private:
+   uint64_t fState[9];     ///< State of the generator
+   unsigned fPosition = 0; ///< Current position in bits
+
+   static constexpr const uint64_t *kA = RanluxppData<p>::kA;
+   static constexpr unsigned kMaxPos = 9 * 64;
+
+   /// Produce next block of random bits
+   void Advance()
+   {
+      mulmod(kA, fState);
+      fPosition = 0;
+   }
+
+public:
+   /// Return the next random bits, generate a new block if necessary
+   uint64_t NextRandomBits()
+   {
+      if (fPosition + w > kMaxPos) {
+         Advance();
+      }
+
+      unsigned idx = fPosition / 64;
+      unsigned offset = fPosition % 64;
+      unsigned numBits = 64 - offset;
```

Comment:
https://wesmckinney.com/blog/avoid-unsigned-integers/ is a nice summary of the concerns.

Reply:
I've changed the code to `int`, added some `assert`s and a cast in `n < left` below to avoid a compiler warning. I'd like to keep the `unsigned` for the carry bits in `mulmod.h` to reduce the number of casts in conjunction with `uint64_t`.